### ISSUE TITLE
config.min_version doesn't work if the new TOML file breaks the old parser

### ISSUE THREAD

**Issue by roblabla** [sagiegurari/cargo-make#267]
**Describe the bug**
cargo-make tries to parse the Makefile.toml entirely before checking the config.min_version key. This causes an issue when the new format cannot be parsed properly by the old parser: we get an error before min_version has a chance to get seen.

**To Reproduce**
Under cargo-make 0.21:

```
[config]
min_version = "0.22.0"

[tasks.bootstrap]
workspace = false
description = "Compiles the i386 bootstrap for debug"
dependencies = ["bootstrap-linker", "install-rust-src"]
command = "cargo"
args = ["xbuild", "--target=i386-unknown-none", "--package=sunrise-bootstrap" ]
```

**Error Stack**
```console
thread 'main' panicked at 'Unable to parse external descriptor, data did not match any variant of untagged enum InstallCrate for key `tasks.bootstrap.install_crate`', C:\Users\felic\.cargo\registry\src\github.com-1ecc6299db9ec823\cargo-make-0.21.0\src/lib\descriptor.rs:228:27
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

**Comment by sagiegurari** [comment]
the min version is used to prevent cargo make running the flow in case the makefile structure is legit but in reality cargo make can't support the needed features.
For example, functions:

```toml
[tasks.split-example]
command = "echo"
args = ["@@split(ENV_VAR,|)"]
```

if you run an old cargo-make, it will work but it will just pass that string as an arg to the command instead of invoking the split function.
in order to prevent such inconsistent behavior, I introduced the min version.
in other words, anything that can break the build (min version or structure change) and prevents cargo-make from running tasks that it doesn't really know how to, seems ok to me as the end result is the same.

**Comment by roblabla** [comment]
To be clear, this is the expected output:

```
[roblabla@roblab kfs]$ cargo make qemu
[cargo-make] INFO - cargo make 0.20.0
[cargo-make] INFO - Using Build File: Makefile.toml
[cargo-make] INFO - Task: qemu
[cargo-make] INFO - Profile: development
[cargo-make] ERROR - cargo make version: 0.20.0 does not meet minimum required version: 0.22.0
[cargo-make] WARN - Build Failed.
```

I'm specifically after this kind of output because it would let my teammates know immediately what's up and what they have to do to get the build working again.

The TOML could get deserialized to a super minimal struct that just looks for `config.min_version`'s value, and emit the error on failure, or something like that.

**Comment by sagiegurari** [comment]
I could load only the initial makefile, but that might not be accurate.
if i want to do it right, it could be costly and complex.
i'll have to load the entire hierarchy of makefiles, some which might require loading scripts to fetch as that min_version might be defined in any one of them.

**Comment by sagiegurari** [comment]
thinking about adding a [global config](https://github.com/sagiegurari/cargo-make#cargo-make-global-config) option to allow this shallow loading of only first makefile just to check min_version without loading extended makefiles.
that would mean you would have to add something in your global config.toml file:

```toml
min_version_shallow_validation = true
```

so this feature would have to be manually opt-in outside your project.
makes sense?

**Comment by roblabla** [comment]
That would work-ish, but I'm admittedly not a huge fan of having people write something in their local config, given I mostly want this to facilitate on-boarding.

I looked around in the codebase. The min_version is checked [right after descriptor loading](https://github.com/sagiegurari/cargo-make/blob/f3fdc37c4a0e667d9b1c3fd96bfb4470a7e163fd/src/lib/cli.rs#L93). Descriptor loading works by calling [`load_external_descriptor`](https://github.com/sagiegurari/cargo-make/blob/b1c42e251e813800d74d727eafcceb00de2468fc/src/lib/descriptor.rs#L192), which will recursively call itself with all the extended makefiles.

So I made https://github.com/roblabla/cargo-make/commit/cc990ed346fb92ea76dcd2dd744ddcecb214e870, which is admittedly not a pretty fix. Basically, `load_descriptor` (and its parents) will return an `Err(String)` containing the min_version if it comes across a Makefile.toml with an incompatible version. It does so before doing any parsing to `cargo-make` data structures. I checked it against my Makefile.toml above, and the result is the one intended. I believe a `min_version` tucked away in a deep Makefile extend would work just as well.

It probably has a few edge-cases (I can think of one: if `min_version` is not a string, it assumes it's not present) but I think it presents a more complete solution that will mesh usefully with #266 (we wouldn't want safe update to fail because the descriptor can't be parsed with the old version).

Would you accept a PR with this?

**Comment by sagiegurari** [comment]
PRs are the best :)
